core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -513,11 +469,116 @@ void toSerializable(DistributedStorage storage)
     parallelExecutor.validate(tasks, getId());
   }
 
-  private TableMetadata getTableMetadata(Scan scan) throws ExecutionException {
-    TransactionTableMetadata metadata = tableMetadataManager.getTransactionTableMetadata(scan);
+  private void validateScanResults(
+      DistributedStorage storage,
+      Scan scan,
+      Map<Key, TransactionResult> results,
+      boolean scanWithLimit)
+      throws ExecutionException, ValidationConflictException {
+    Scanner scanner = null;
+    try {
+      // Only get tx_id, tx_version and primary key columns because we use only them to compare
+      scan.clearProjections();
+      scan.withProjection(Attribute.ID).withProjection(Attribute.VERSION);
+      ScalarDbUtils.addProjectionsForKeys(scan, getTableMetadata(scan));
+
+      if (!scanWithLimit) {
+        scanner = storage.scan(scan);
+      } else {
+        // Get a scanner without the limit if the scan has a limit
+        scanner = storage.scan(Scan.newBuilder(scan).limit(0).build());
+      }
+
+      Iterator<Entry<Key, TransactionResult>> resultsIterator = results.entrySet().iterator();
+
+      Optional<Result> latestResult = scanner.one();
+
+      Entry<Key, TransactionResult> resultsEntry = null;
+      if (resultsIterator.hasNext()) {
+        resultsEntry = resultsIterator.next();
+      }
+
+      while (latestResult.isPresent() && resultsEntry != null) {
+        TransactionResult latestTransactionResult = new TransactionResult(latestResult.get());
+        Key key = new Key(scan, latestTransactionResult);
+
+        // Ignore records that this transaction has prepared (and that are in the write set or
+        // delete set)
+        if (latestTransactionResult.getId() != null && latestTransactionResult.getId().equals(id)) {
+          latestResult = scanner.one();
+
+          if (resultsEntry.getKey().equals(key)) {
```

Comment:
I might be wrong, but this implementation assumes `resultsIterator` and `scanner` return values in the same order? I'm not sure why `results` is not used as a Map as is...

Reply:
> this implementation assumes resultsIterator and scanner return values in the same order?

Yes. The implementation of the map `results` is `LinkedHashMap`, so the order is preserved.

This logic assumes that the scanner for the same scan returns records in the same order. What do you think?

Reply:
Glad to hear that `LinkedHashMap` here. It would be great if the parameter type is explicitly `LinkedHashMap`.

As for the underlying database/storage, I don't think they always return scanned records in the same order. They seem to often return records in the same order, but it depends on temporary physical data structure that can change.

Reply:
For partition scans, the storage abstraction guarantees that records are returned in clustering order, so I think it’s fine in that case. For cross-partition scans and index-based scans, there’s no such ordering guarantee. However, in most cases, the records seem to be returned in a consistent order, so I assumed we could rely on that.

That said, if we can’t assume consistent ordering, it would complicate things quite a bit. Let me discuss this. Thanks!